File: app/infrastructure/encryption/layer_header.py
"""``layer.header`` — the only unencrypted file in a private layer.

It holds the *wrapped* layer data key, never the key itself:

    password ──Argon2id──▶ KEK_pw ──AEAD-wrap──▶ [LDK]
    recovery key ─────────▶ KEK_rk ──AEAD-wrap──▶ [LDK]   (optional, same LDK)

Consequences of wrapping rather than deriving:

* **Changing the password is cheap.** Re-derive a KEK, rewrap the same LDK. No
  object is touched, so a password change on a 10 GB layer is instant.
* **Rotating the key is possible at all.** A new LDK means every object must be
  re-encrypted — expensive, but it is the only thing that actually revokes a
  collaborator who kept a copy of the old key.
* **A recovery key is not a backdoor.** It wraps the same LDK independently, so it
  grants exactly what the password grants, and nothing more. Losing both makes the
  layer unrecoverable — there is no third door, and we say so.

The header deliberately contains no title, no object count, and no hint about the
contents. What it does reveal, unavoidably: that this is a Strata private layer,
when it was created and last touched, and the KDF parameters.
"""

from __future__ import annotations

import json
import secrets
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Final

from app.infrastructure.encryption.primitives import (
    KEY_BYTES,
    DecryptionError,
    KdfParams,
    decrypt,
    derive_key,
    encrypt,
    random_key,
)

HEADER_FILENAME: Final = "layer.header"
HEADER_FORMAT_VERSION: Final = 1

# AAD for the wrap. Binds the envelope to its purpose, so a password envelope can
# never be presented as a recovery envelope (or vice versa) to confuse the KDF.
_AAD_PASSWORD: Final = b"strata:layer-key:password:v1"
_AAD_RECOVERY: Final = b"strata:layer-key:recovery:v1"

RECOVERY_KEY_BYTES: Final = 32
_RECOVERY_ALPHABET: Final = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"  # no I/O/0/1


def generate_recovery_key() -> str:
    """A 256-bit recovery key, grouped for transcription.

    Crockford-ish alphabet: no I, O, 0 or 1, because this is a string a human will
    write on paper and type back in a year, probably in a hurry.
    """
    raw = secrets.token_bytes(RECOVERY_KEY_BYTES)
    value = int.from_bytes(raw, "big")
    digits: list[str] = []
    while value:
        value, remainder = divmod(value, 32)
        digits.append(_RECOVERY_ALPHABET[remainder])
    text = "".join(reversed(digits)).rjust(52, _RECOVERY_ALPHABET[0])
    return "-".join(text[i : i + 4] for i in range(0, len(text), 4))


def normalise_recovery_key(value: str) -> str:
    return value.replace("-", "").replace(" ", "").strip().upper()


@dataclass
class WrappedKey:
    """One envelope around the layer data key."""

    kdf: KdfParams
    nonce: bytes
    ciphertext: bytes

    def to_json(self) -> dict[str, Any]:
        return {
            "kdf": self.kdf.to_json(),
            "nonce": self.nonce.hex(),
            "ciphertext": self.ciphertext.hex(),
        }

    @classmethod
    def from_json(cls, raw: dict[str, Any]) -> WrappedKey:
        return cls(
            kdf=KdfParams.from_json(raw["kdf"]),
            nonce=bytes.fromhex(raw["nonce"]),
            ciphertext=bytes.fromhex(raw["ciphertext"]),
        )


def wrap_key(secret: str, layer_key: bytes, *, aad: bytes) -> WrappedKey:
    params = KdfParams.new()
    kek = derive_key(secret, params)
    nonce, ciphertext = encrypt(kek, layer_key, aad)
    return WrappedKey(kdf=params, nonce=nonce, ciphertext=ciphertext)


def unwrap_key(secret: str, wrapped: WrappedKey, *, aad: bytes) -> bytes:
    kek = derive_key(secret, wrapped.kdf)
    key = decrypt(kek, wrapped.nonce, wrapped.ciphertext, aad)
    if len(key) != KEY_BYTES:
        raise DecryptionError()
    return key


@dataclass
class LayerHeader:
    format_version: int = HEADER_FORMAT_VERSION
    layer_id: str = ""
    created_at: str = ""
    updated_at: str = ""
    key_generation: int = 1
    manifest_object_id: str = ""
    password_envelope: WrappedKey | None = None
    recovery_envelope: WrappedKey | None = None
    padding_enabled: bool = True
    extra: dict[str, Any] = field(default_factory=dict)

    # -- key operations ------------------------------------------------------

    @classmethod
    def create(
        cls,
        *,
        layer_id: str,
        password: str,
        created_at: str,
        recovery_key: str | None = None,
    ) -> tuple[LayerHeader, bytes]:
        """Create a header around a *fresh random* layer key.

        The password never becomes the key. It wraps it.
        """
        layer_key = random_key()
        header = cls(
            layer_id=layer_id,
            created_at=created_at,
            updated_at=created_at,
            password_envelope=wrap_key(password, layer_key, aad=_AAD_PASSWORD),
        )
        if recovery_key:
            header.recovery_envelope = wrap_key(
                normalise_recovery_key(recovery_key), layer_key, aad=_AAD_RECOVERY
            )
        return header, layer_key

    def unlock_with_password(self, password: str) -> bytes:
        if self.password_envelope is None:
            raise DecryptionError()
        return unwrap_key(password, self.password_envelope, aad=_AAD_PASSWORD)

    def unlock_with_recovery_key(self, recovery_key: str) -> bytes:
        if self.recovery_envelope is None:
            raise DecryptionError()
        return unwrap_key(
            normalise_recovery_key(recovery_key), self.recovery_envelope, aad=_AAD_RECOVERY
        )

    def change_password(self, old_password: str, new_password: str) -> None:
        """Rewrap the *same* layer key. No object is re-encrypted.

        This does not revoke anyone who already has the layer key — it only changes
        what unlocks it on this device. Revocation is key *rotation*, and it is a
        different, expensive operation.
        """
        layer_key = self.unlock_with_password(old_password)
        self.password_envelope = wrap_key(new_password, layer_key, aad=_AAD_PASSWORD)

    def set_recovery_key(self, password: str, recovery_key: str) -> None:
        layer_key = self.unlock_with_password(password)
        self.recovery_envelope = wrap_key(
            normalise_recovery_key(recovery_key), layer_key, aad=_AAD_RECOVERY
        )

    def rewrap_for_rotation(self, password: str, new_layer_key: bytes) -> None:
        """Point the envelopes at a new layer key. The caller re-encrypts objects."""
        self.password_envelope = wrap_key(password, new_layer_key, aad=_AAD_PASSWORD)
        # A recovery key that still opened the *old* key would be a hole straight
        # through the rotation, so it is dropped and must be re-issued.
        self.recovery_envelope = None
        self.key_generation += 1

    # -- persistence ---------------------------------------------------------

    def to_json(self) -> dict[str, Any]:
        return {
            "format_version": self.format_version,
            "layer_id": self.layer_id,
            "created_at": self.created_at,
            "updated_at": self.updated_at,
            "key_generation": self.key_generation,
            "manifest_object_id": self.manifest_object_id,
            "padding_enabled": self.padding_enabled,
            "password": self.password_envelope.to_json() if self.password_envelope else None,
            "recovery": self.recovery_envelope.to_json() if self.recovery_envelope else None,
            **self.extra,
        }

    @classmethod
    def from_json(cls, raw: dict[str, Any]) -> LayerHeader:
        version = int(raw.get("format_version", 0))
        if version > HEADER_FORMAT_VERSION:
            raise DecryptionError("This layer was created by a newer version of Strata.")
        return cls(
            format_version=version,
            layer_id=str(raw.get("layer_id", "")),
            created_at=str(raw.get("created_at", "")),
            updated_at=str(raw.get("updated_at", "")),
            key_generation=int(raw.get("key_generation", 1)),
            manifest_object_id=str(raw.get("manifest_object_id", "")),
            padding_enabled=bool(raw.get("padding_enabled", True)),
            password_envelope=(
                WrappedKey.from_json(raw["password"]) if raw.get("password") else None
            ),
            recovery_envelope=(
                WrappedKey.from_json(raw["recovery"]) if raw.get("recovery") else None
            ),
        )

    def save(self, root: Path) -> None:
        root.mkdir(parents=True, exist_ok=True)
        path = root / HEADER_FILENAME
        temporary = path.with_suffix(".header.tmp")
        temporary.write_text(json.dumps(self.to_json(), indent=2), encoding="utf-8")
        # Atomic: losing the header to a half-written file would lose the layer.
        temporary.replace(path)

    @classmethod
    def load(cls, root: Path) -> LayerHeader:
        path = root / HEADER_FILENAME
        if not path.is_file():
            raise DecryptionError("This layer has no header.")
        try:
            raw = json.loads(path.read_text(encoding="utf-8"))
        except (json.JSONDecodeError, OSError) as exc:
            raise DecryptionError("The layer header could not be read.") from exc
        if not isinstance(raw, dict):
            raise DecryptionError("The layer header is not valid.")
        return cls.from_json(raw)

File: dotnet/src/Strata.Infrastructure/Encryption/LayerHeader.cs
using System.Numerics;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using Strata.Core.Encryption;
using Strata.Core.Errors;
using Strata.Infrastructure.Storage;

namespace Strata.Infrastructure.Encryption;

/// <summary>One envelope around the layer data key.</summary>
public sealed record WrappedKey(KdfParams Kdf, byte[] Nonce, byte[] Ciphertext)
{
    public JsonObject ToJson() => new()
    {
        ["kdf"] = Kdf.ToJson(),
        ["nonce"] = Convert.ToHexString(Nonce).ToLowerInvariant(),
        ["ciphertext"] = Convert.ToHexString(Ciphertext).ToLowerInvariant(),
    };

    public static WrappedKey FromJson(JsonObject raw)
    {
        try
        {
            return new WrappedKey(
                KdfParams.FromJson(raw["kdf"]!.AsObject()),
                Convert.FromHexString(raw["nonce"]!.GetValue<string>()),
                Convert.FromHexString(raw["ciphertext"]!.GetValue<string>()));
        }
        catch (Exception exc) when (exc is NullReferenceException or InvalidOperationException
            or FormatException or ArgumentException)
        {
            throw new DecryptionException("The layer header is not valid.", exc);
        }
    }
}

/// <summary>
/// <c>layer.header</c> — the only unencrypted file in a private layer. Port of
/// <c>app/infrastructure/encryption/layer_header.py</c>.
/// </summary>
/// <remarks>
/// <para>It holds the <em>wrapped</em> layer data key, never the key itself:</para>
/// <code>
/// password     ──Argon2id──▶ KEK_pw ──AEAD-wrap──▶ [LDK]
/// recovery key ─────────────▶ KEK_rk ──AEAD-wrap──▶ [LDK]   (optional, same LDK)
/// </code>
/// <para>
/// Wrapping rather than deriving makes a password change cheap (rewrap, touch no
/// object), makes key rotation possible at all, and keeps a recovery key from being
/// a backdoor — it wraps the same LDK independently, so it grants exactly what the
/// password grants and nothing more. Losing both makes the layer unrecoverable;
/// there is no third door.
/// </para>
/// </remarks>
public sealed class LayerHeader
{
    public const string HeaderFilename = "layer.header";
    public const int HeaderFormatVersion = 1;

    /// <summary>AAD binding the wrap to its purpose, so a password envelope can never
    /// be presented as a recovery envelope (or vice versa) to confuse the KDF.</summary>
    public static ReadOnlySpan<byte> AadPassword => "strata:layer-key:password:v1"u8;

    public static ReadOnlySpan<byte> AadRecovery => "strata:layer-key:recovery:v1"u8;

    public const int RecoveryKeyBytes = 32;

    /// <summary>Crockford-ish: no I, O, 0 or 1 — this is written on paper and typed back.</summary>
    private const string RecoveryAlphabet = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";

    private const int RecoveryDigits = 52;

    public int FormatVersion { get; set; } = HeaderFormatVersion;

    public string LayerId { get; set; } = string.Empty;

    public string CreatedAt { get; set; } = string.Empty;

    public string UpdatedAt { get; set; } = string.Empty;

    public int KeyGeneration { get; set; } = 1;

    public string ManifestObjectId { get; set; } = string.Empty;

    public WrappedKey? PasswordEnvelope { get; set; }

    public WrappedKey? RecoveryEnvelope { get; set; }

    public bool PaddingEnabled { get; set; } = true;

    // -- recovery keys -------------------------------------------------------

    /// <summary>A 256-bit recovery key, grouped for transcription.</summary>
    public static string GenerateRecoveryKey()
    {
        var raw = RandomNumberGenerator.GetBytes(RecoveryKeyBytes);
        var value = new BigInteger(raw, isUnsigned: true, isBigEndian: true);
        var digits = new StringBuilder();
        while (value > BigInteger.Zero)
        {
            value = BigInteger.DivRem(value, 32, out var remainder);
            digits.Append(RecoveryAlphabet[(int)remainder]);
        }

        var text = new string(digits.ToString().Reverse().ToArray())
            .PadLeft(RecoveryDigits, RecoveryAlphabet[0]);

        var grouped = new StringBuilder();
        for (var i = 0; i < text.Length; i += 4)
        {
            if (i > 0)
            {
                grouped.Append('-');
            }

            grouped.Append(text.AsSpan(i, Math.Min(4, text.Length - i)));
        }

        return grouped.ToString();
    }

    public static string NormaliseRecoveryKey(string value)
        => value.Replace("-", string.Empty).Replace(" ", string.Empty).Trim().ToUpperInvariant();

    // -- key operations ------------------------------------------------------

    public static WrappedKey WrapKey(string secret, ReadOnlySpan<byte> layerKey, ReadOnlySpan<byte> aad)
    {
        var parameters = KdfParams.New();
        var kek = AeadPrimitives.DeriveKey(secret, parameters);
        try
        {
            var (nonce, ciphertext) = AeadPrimitives.Encrypt(kek, layerKey, aad);
            return new WrappedKey(parameters, nonce, ciphertext);
        }
        finally
        {
            AeadPrimitives.Zeroize(kek);
        }
    }

    public static byte[] UnwrapKey(string secret, WrappedKey wrapped, ReadOnlySpan<byte> aad)
    {
        var kek = AeadPrimitives.DeriveKey(secret, wrapped.Kdf);
        try
        {
            var key = AeadPrimitives.Decrypt(kek, wrapped.Nonce, wrapped.Ciphertext, aad);
            if (key.Length != CryptoConstants.KeyBytes)
            {
                throw new DecryptionException();
            }

            return key;
        }
        finally
        {
            AeadPrimitives.Zeroize(kek);
        }
    }

    /// <summary>Create a header around a <em>fresh random</em> layer key.</summary>
    /// <remarks>The password never becomes the key. It wraps it.</remarks>
    public static (LayerHeader Header, byte[] LayerKey) Create(
        string layerId,
        string password,
        string createdAt,
        string? recoveryKey = null)
    {
        var layerKey = AeadPrimitives.RandomKey();
        var header = new LayerHeader
        {
            LayerId = layerId,
            CreatedAt = createdAt,
            UpdatedAt = createdAt,
            PasswordEnvelope = WrapKey(password, layerKey, AadPassword),
        };

        if (!string.IsNullOrEmpty(recoveryKey))
        {
            header.RecoveryEnvelope = WrapKey(NormaliseRecoveryKey(recoveryKey), layerKey, AadRecovery);
        }

        return (header, layerKey);
    }

    public byte[] UnlockWithPassword(string password)
    {
        if (PasswordEnvelope is null)
        {
            throw new DecryptionException();
        }

        return UnwrapKey(password, PasswordEnvelope, AadPassword);
    }

    public byte[] UnlockWithRecoveryKey(string recoveryKey)
    {
        if (RecoveryEnvelope is null)
        {
            throw new DecryptionException();
        }

        return UnwrapKey(NormaliseRecoveryKey(recoveryKey), RecoveryEnvelope, AadRecovery);
    }

    /// <summary>Rewrap the <em>same</em> layer key. No object is re-encrypted.</summary>
    /// <remarks>
    /// This does not revoke anyone who already has the layer key — it only changes
    /// what unlocks it on this device. Revocation is key <em>rotation</em>, and it is
    /// a different, expensive operation.
    /// </remarks>
    public void ChangePassword(string oldPassword, string newPassword)
    {
        var layerKey = UnlockWithPassword(oldPassword);
        try
        {
            PasswordEnvelope = WrapKey(newPassword, layerKey, AadPassword);
        }
        finally
        {
            AeadPrimitives.Zeroize(layerKey);
        }
    }

    public void SetRecoveryKey(string password, string recoveryKey)
    {
        var layerKey = UnlockWithPassword(password);
        try
        {
            RecoveryEnvelope = WrapKey(NormaliseRecoveryKey(recoveryKey), layerKey, AadRecovery);
        }
        finally
        {
            AeadPrimitives.Zeroize(layerKey);
        }
    }

    /// <summary>Point the envelopes at a new layer key. The caller re-encrypts objects.</summary>
    public void RewrapForRotation(string password, ReadOnlySpan<byte> newLayerKey)
    {
        PasswordEnvelope = WrapKey(password, newLayerKey, AadPassword);

        // A recovery key that still opened the *old* key would be a hole straight
        // through the rotation, so it is dropped and must be re-issued.
        RecoveryEnvelope = null;
        KeyGeneration += 1;
    }

    // -- persistence ---------------------------------------------------------

    public JsonObject ToJson() => new()
    {
        ["format_version"] = FormatVersion,
        ["layer_id"] = LayerId,
        ["created_at"] = CreatedAt,
        ["updated_at"] = UpdatedAt,
        ["key_generation"] = KeyGeneration,
        ["manifest_object_id"] = ManifestObjectId,
        ["padding_enabled"] = PaddingEnabled,
        ["password"] = PasswordEnvelope?.ToJson(),
        ["recovery"] = RecoveryEnvelope?.ToJson(),
    };

    public static LayerHeader FromJson(JsonObject raw)
    {
        var version = raw["format_version"]?.GetValue<int>() ?? 0;
        if (version > HeaderFormatVersion)
        {
            throw new DecryptionException("This layer was created by a newer version of Strata.");
        }

        return new LayerHeader
        {
            FormatVersion = version,
            LayerId = raw["layer_id"]?.GetValue<string>() ?? string.Empty,
            CreatedAt = raw["created_at"]?.GetValue<string>() ?? string.Empty,
            UpdatedAt = raw["updated_at"]?.GetValue<string>() ?? string.Empty,
            KeyGeneration = raw["key_generation"]?.GetValue<int>() ?? 1,
            ManifestObjectId = raw["manifest_object_id"]?.GetValue<string>() ?? string.Empty,
            PaddingEnabled = raw["padding_enabled"]?.GetValue<bool>() ?? true,
            PasswordEnvelope = raw["password"] is JsonObject password ? WrappedKey.FromJson(password) : null,
            RecoveryEnvelope = raw["recovery"] is JsonObject recovery ? WrappedKey.FromJson(recovery) : null,
        };
    }

    public void Save(string root)
    {
        Directory.CreateDirectory(root);
        var json = ToJson().ToJsonString(new JsonSerializerOptions { WriteIndented = true });
        // Atomic: losing the header to a half-written file would lose the layer.
        AtomicFile.WriteText(Path.Combine(root, HeaderFilename), json);
    }

    public static LayerHeader Load(string root)
    {
        var path = Path.Combine(root, HeaderFilename);
        if (!File.Exists(path))
        {
            throw new DecryptionException("This layer has no header.");
        }

        JsonNode? node;
        try
        {
            node = JsonNode.Parse(File.ReadAllText(path));
        }
        catch (Exception exc) when (exc is JsonException or IOException or UnauthorizedAccessException)
        {
            throw new DecryptionException("The layer header could not be read.", exc);
        }

        if (node is not JsonObject raw)
        {
            throw new DecryptionException("The layer header is not valid.");
        }

        return FromJson(raw);
    }
}
